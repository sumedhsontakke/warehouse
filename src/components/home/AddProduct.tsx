import { memo } from "react";
import { useNavigate } from "react-router-dom";
import styled from "styled-components"

export const AddProductWrapper = styled.div`
    font-size: 18px;
    font-weight: 600;
    cursor: pointer;
    margin-top: 30px;
    text-align: right;
    :hover{
        color: var(--bs-link-color);
    }
`;

function AddProduct(){
    const navigate = useNavigate();
    return(
        <AddProductWrapper onClick={()=> navigate("/add-product")}>
            <span>Add Product</span>
        </AddProductWrapper>
    )
}

export default memo(AddProduct)