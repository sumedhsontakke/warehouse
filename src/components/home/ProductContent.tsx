import styled from "styled-components";

export const ProductContent = styled.div`
    width: 30%;
    border: 2px solid #e7e7e7;
    margin-right: 5%;
    cursor: pointer;
    padding: 20px;
    font-weight: 600;
    margin-bottom: 35px;
    box-sizing: border-box;
    display: flex;
    flex-wrap: wrap;
    > span {
        width: 50%
    }
    > span:nth-last-child(1){
        margin-left: auto;
        display: block;
    }
    > div{
        width: 100%;
        margin-top: 30px;
        margin-bottom: 20px;
        :hover{
            color: var(--bs-gray-900);
        }
    }
    :hover{
        color: var(--bs-link-color)
    }
    :nth-child(3n){
        margin-right: 0;
    }
    @media screen and (max-width: 768px){
        width: 100%;
        margin-right: 0;
    }
`;