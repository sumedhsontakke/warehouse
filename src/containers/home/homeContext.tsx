import { createContext, useContext } from 'react';
import { IProductsContext } from '../../misc/interfaces';

export const ProductsContext = createContext<IProductsContext>({
    products: null,
    setProducts: () => {},
    saveProductList: () =>{}
});

export const useProductsContext = () => useContext(ProductsContext);