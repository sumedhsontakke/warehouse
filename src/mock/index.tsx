import { AxiosResponse } from "axios";
import { IProduct } from "../misc/interfaces";
export const product: IProduct[] = [
    {
      "id": "a269a247-0d38-4b47-9630-79c9ae545b68",
      "name": "Dining Chair",
      "requiredArticles": [
        {
          "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
          "amountRequired": 4
        },
        {
          "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
          "amountRequired": 1
        },
        {
          "id": "addc65a8-c759-41d8-a18a-89fe446ad484",
          "amountRequired": 8
        }
      ]
    }
  ];


export const mockresponse: AxiosResponse = {headers: {}, config: {}, statusText: "", status: 100, data: product};

export const salesMockResponse: AxiosResponse = {headers: {}, config: {}, statusText: "", status: 100,
     data:[
        {
          "id": "da5846e7-ba53-4189-b062-81fc64582431",
          "createdAt": "2020-11-13T12:00:00.123",
          "productId": "a269a247-0d38-4b47-9630-79c9ae545b68",
          "amountSold": 1
        }
     ] 
};

export const articleMockResponse: AxiosResponse = {headers: {}, config: {}, statusText: "", status: 100,
    data:[
        {
            "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
            "name": "Leg",
            "amountInStock": 12
        },
    ] 
};

export const  SalesProps = {
    salesDetail: [
        {
            "id": "da5846e7-ba53-4189-b062-81fc64582431",
            "createdAt": "2020-11-13T12:00:00.123",
            "productId": "a269a247-0d38-4b47-9630-79c9ae545b68",
            "amountSold": 1,
            "name": "test"
          },
    ],
    isLoading: false,
    productList: null
}

export const articleComponentProps = {
    articles: [
        {
            id: "testID",
            name: "article name",
            amountInStock: 5,  
            amountToSubtract: 2,
            amountRequired: 4
        }
    ],
    updateArticle: () =>{},
    setArticles: () => {}
}

export const productDetailsProps = {
    "id": "a269a247-0d38-4b47-9630-79c9ae545b68",
    "name": "Dining Chair",
    "articles": [
        {
            "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
            "amountRequired": 4,
            "amountInStock": 2,
            "name": "article 1"
        },
        {
            "id": "831b92b8-677b-42cc-a585-335ea4ccccb6",
            "amountRequired": 5,
            "amountInStock": 1,
            "name": "article 2"
        },
    ],
    "requiredArticles": [
        {
          "id": "0517f083-0e15-4876-8d1f-6fa45900431c",
          "amountRequired": 4
        },
    ],
    saleProduct : () =>{
        
    }
        
}

export const mockForValidate = {
    product: "test",
    requiredArticles: [{"id":"0517f083-0e15-4876-8d1f-6fa45900431c","amountRequired":12}],
    allArticles: [{"id":"0517f083-0e15-4876-8d1f-6fa45900431c","name":"Patched Leg","amountInStock":12},{"id":"addc65a8-c759-41d8-a18a-89fe446ad484","name":"Screw","amountInStock":41},{"id":"831b92b8-677b-42cc-a585-335ea4ccccb6","name":"Seat","amountInStock":8},{"id":"6892b98b-9b87-4520-9a9e-7528f1d78cb4","name":"Table Top","amountInStock":12},{"id":"41a84cdc-c70b-4924-9907-319572a45ca3","name":"Leg","amountInStock":12}]
}