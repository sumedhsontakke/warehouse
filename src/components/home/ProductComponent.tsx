import { memo } from "react"
import { IProduct } from '../../misc/interfaces';
import { ProductContent } from './ProductContent';
import { useNavigate } from "react-router-dom";

function ProductComponent({
    product,
}: {product: IProduct}){
    const navigate = useNavigate();
    return(
        <ProductContent
            onClick={() => navigate(`/product-details/${product.id}`)}
        >
            {product.name}
        </ProductContent>
    )
}

export default memo(ProductComponent);