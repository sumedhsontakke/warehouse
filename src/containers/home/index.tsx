import { useEffect, useState } from 'react';
import { IProduct } from '../../misc/interfaces';
import { getProductList } from "../../misc/request";
import { ProductsContext } from "./homeContext";
import FallBackUI from '../../components/FallBackUI';
import HomeComponent from '../../components/home';
import SpinnerComponent from '../../components/common/SpinnerComponent';

function Home(){

    const [products, setProducts] = useState<IProduct[] | null>(null);
    const [isSpinner, setSpinner] = useState<Boolean>(false);
    const [error, setError] = useState<Boolean>(false);
    
    useEffect(() =>{
        console.log("home container rerender", products);
        if(!products)
            saveProductList(1); 
    }, [])

    const saveProductList = (retry: number) => { 
        setSpinner(true);
        if(retry <= 2){
            try{
                const result = getProductList();
                result.then(resp =>{ 
                    if(resp.status === 200){
                        setProducts(resp.data);
                    }
                    setSpinner(false);
                })
                .catch((error) =>{
                    setSpinner(false);
                    saveProductList(retry + 1);
                })
            } catch (e){
                console.log("Failed while fetching product list");
                setSpinner(false);
                retry < 2 && saveProductList(retry + 1);
            }
        } else {
            setSpinner(false);
            setError(true);
        }
    }

    return(
        <div>
            <ProductsContext.Provider value={{products, setProducts, saveProductList}}>
                { !error && <HomeComponent /> }
                { isSpinner && <SpinnerComponent /> }
                { error && <FallBackUI /> }
            </ProductsContext.Provider>
        </div>
    )
}

export default Home;