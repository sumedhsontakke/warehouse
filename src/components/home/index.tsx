import React, { useEffect } from 'react';
import { useProductsContext } from '../../containers/home/homeContext';

import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

import HeaderComponent from '../common/HeaderComponent';
import { ProductWrapper } from '../common/commonStyle';
import ProductComponent from './ProductComponent';
import { IProduct } from '../../misc/interfaces';
import AddProduct from './AddProduct';

function HomeComponent(){

    const { products } = useProductsContext()

    useEffect(() =>{
        console.log("home component rerender");
    }, [])

    return(
        <div>
            <HeaderComponent />
            <Container>
                <Row>
                    <Col>
                        <AddProduct />
                        <ProductWrapper>
                            {
                                products ?
                                products.map((product:IProduct, index) => <ProductComponent key={index} product={product} /> )
                                : ""
                            }
                        </ProductWrapper>
                    </Col>
                </Row>
            </Container>
        </div>
    )
}

export default HomeComponent;