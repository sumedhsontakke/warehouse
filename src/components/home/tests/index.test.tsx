
import { render, screen } from "@testing-library/react";
import HomeComponent from "../index";
import {BrowserRouter as Router} from 'react-router-dom';
import * as content from "../../../containers/home/homeContext";
import { act } from 'react-dom/test-utils';
import { mockresponse, product } from '../../../mock';

describe('<HomeComponent />' , () =>{        
    it('should render and match the snapshot', async() => {
      const mock = jest.spyOn(content, "useProductsContext")
      .mockReturnValue({products: product, setProducts : ()=> {}, saveProductList : ()=> {}});
        const {
          container: { firstChild },
        } = 
        render(
            <Router>
                <HomeComponent />
            </Router>
        );
        await act(async () => {
          expect(firstChild).toMatchSnapshot();
          mock.mockRestore();
        });  
    });  

    it('should render product name', async() => {
        const mock = jest.spyOn(content, "useProductsContext")
        .mockReturnValue({products: product, setProducts : ()=> {}, saveProductList : ()=> {}});
       
          const {
            container: { firstChild },
          } = 
          render(
              <Router>
                  <HomeComponent />
              </Router>
          );
          await act(async () => {
            expect(await screen.findByText(product[0].name)).toBeInTheDocument();
            mock.mockRestore();
          });  
      });  
})