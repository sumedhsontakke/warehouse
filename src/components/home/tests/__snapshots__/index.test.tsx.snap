// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeComponent /> should render and match the snapshot 1`] = `
<div>
  <header
    class="sc-gsnTZi drncPl"
  >
    <div
      class="container"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <img
            alt="logo"
            class="sc-bczRLJ RuITL"
            src="sample-logo.png"
          />
        </div>
        <div
          class="col"
        >
          <nav
            class="sc-dkzDqf dEMIWt"
          >
            <a
              href="/"
            >
              Home
            </a>
            <a
              href="/sales"
            >
              Sales
            </a>
            <a
              href="/articles"
            >
              Articles
            </a>
          </nav>
        </div>
      </div>
    </div>
  </header>
  <div
    class="container"
  >
    <div
      class="row"
    >
      <div
        class="col"
      >
        <div
          class="sc-jSMfEi fqlLpr"
        >
          <span>
            Add Product
          </span>
        </div>
        <div
          class="sc-hKMtZM cIMpSZ"
        >
          <div
            class="sc-eCYdqJ byMvSi"
          >
            Dining Chair
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
